File: 53-maximum-subarray/maximum-subarray.java
class Solution {
    // Basic Array Iteration - TC: O(n), SC: O(1)
    public int maxSubArray(int[] nums) {
        int result = nums[0], currSum = 0;

        for (int n : nums) {
            // If currSum becomes negative, discard its value
            if (currSum < 0) currSum = 0;
            
            currSum += n;
            result = Math.max(result, currSum);
        }

        return result;
    }
}

File: 53-maximum-subarray/maximum-subarray.py
class Solution:
    # Kadane's Algorithm - TC: O(n), SC: O(1)
    def maxSubArray(self, nums: List[int]) -> int:
        maxSum = nums[0]
        currSum = 0

        for n in nums:
            currSum = max(n, currSum + n)
            maxSum = max(currSum, maxSum)

        return maxSum
    
    # Brute Force Solution - TC: O(n^2), SC: O(1)
    def maxSubArray2(self, nums: List[int]) -> int:
        maxSum = nums[0]

        for i in range(len(nums)):
            currSum = nums[i]
            for j in range(i+1, len(nums)):
                currSum += nums[j]
                maxSum = max(maxSum, currSum)
        
        return maxSum